File: Leetcode/PY/Leetcode400-第N位数字.py
class Solution:
    def findNthDigit(self, n: int) -> int:
        idx,base,s = 1,9,1
        while n > base * idx:
            n -= base * idx
            base *= 10
            s *= 10
            idx += 1
        s += (n + idx - 1) // idx - 1
        n = n % idx if n % idx != 0 else idx
        return ord(str(s)[n - 1]) - ord('0')

File: Leetcode/Rust/Leetcode400-第N位数字.rs
impl Solution {
    pub fn find_nth_digit(mut n: i32) -> i32 {
        let mut n = n as i64;
        let (mut cnt,mut base,mut s) = (1,9 as i64,1 as i64);
        while n > base * cnt{
            n -= base * cnt;
            cnt += 1;
            base *= 10;
            s *= 10;
        }
        s += (n + cnt - 1) / cnt - 1;
        if n % cnt == 0 {
            n = cnt;
        } else{
            n = n % cnt;
        }
        (s.to_string().as_bytes()[n as usize - 1] - b'0') as i32
    }
}